media/multipart/src/main/java/org/glassfish/jersey/media/multipart/ContentDisposition.java

```diff
@@ -192,6 +207,47 @@ private void createParameters() throws ParseException {
         size = createLong("size");
     }
 
+    private String defineFileName() throws ParseException {
+
+        final String fileName = parameters.get("filename");
+        final String fileNameExt = parameters.get("filename*");
+
+        if (fileNameExt == null) {
+            return fileName;
+        }
+
+        final Matcher matcher = FILENAME_EXT_VALUE_PATTERN.matcher(fileNameExt);
+
+        if (matcher.matches()) {
+
+            final String fileNameValueChars = matcher.group(FILENAME_GROUP_NAME);
+            if (isFilenameValueCharsEncoded(fileNameValueChars)) {
+                return fileNameExt;
+            }
+
+            if (matcher.group(CHARSET_GROUP_NAME).equalsIgnoreCase("UTF-8")) {
+                return new StringBuilder(matcher.group(CHARSET_GROUP_NAME))
+                        .append("'")
+                        .append(matcher.group(LANG_GROUP_NAME) == null ? "" : matcher.group(LANG_GROUP_NAME))
```

Comment:
Regexps are possibly not the fastest piece of code. Can you store the result of `matcher.group(LANG_GROUP_NAME)` and reuse?

Reply:
Done.